Add scaleway provider

raskyld opened the pull request:
I've started to draft a Scaleway provider, the first commit contains only the API and the links between the CLI flags and the said API.

I open it so you can start suggestions / remarks :)

raskyld commented:
@anbraten Is it planned to use a config file instead of CLI Flags and Env Vars for users to describe their agent pools?

raskyld commented:
We should discuss about it because in terms of UX, if we want the auto-scaler to manage different agent pools (with different labels set) this will be really heavy on environment variables and a structure file format like YAML or TOML would be better IMHO.

As a middle-ground maybe we could use `godotenv` with an environment variable `WOODPECKER_AUTOSCALER_AGENT_POOLS_PATH` that loads such a file 🤷

anbraten commented:
> We should discuss about it because in terms of UX, if we want the auto-scaler to manage different agent pools (with different labels set) this will be really heavy on environment variables and a structured file format like YAML or TOML would be better IMHO.

Currently the design is to use one autoscaler per agent-pool. Its a pretty lightweight tool and makes the code / config much simpler. 

In container / deployment setups env vars are normally way easier to setup IMO than mounting a config file, so wouldn't per se agree that a config file is better suited.

raskyld commented:
Alright, I didn't have the argument that we will use multiple autoscaler per pool, that's a pretty neat design :)

raskyld commented:
I am going to comply with the CI checks just before the MR leave the draft state btw

anbraten commented:
> Alright, I didn't have the argument that we will use multiple autoscaler per pool, that's a pretty neat design :)

It might be a consideration in the long-run to not have 100 (lets first get to such an amount 😅) agents querying the server for queue information, but til then I think it ways easier to have separate ones.

raskyld commented:
I am going to test the PR later in the afternoon, do you have any hint on how to setup a testing environment as I am pretty new to WP? 🙏

raskyld commented:
Solves #51

anbraten commented:
> I am going to test the PR later in the afternoon, do you have any hint on how to setup a testing environment as I am pretty new to WP? 🙏

I normally test using a simple docker-compose with gitea. Sth like this:

```yml
version: '3'

services:
   woodpecker-server:
     image: woodpeckerci/woodpecker-server:next
     ports:
       - 8000:8000
     volumes:
     environment:
       - WOODPECKER_OPEN=true
       - WOODPECKER_ADMIN=anbraten
       - WOODPECKER_HOST=${WOODPECKER_HOST}
       - WOODPECKER_GITEA=true
       - WOODPECKER_GITEA_CLIENT=${WOODPECKER_GITHUB_CLIENT}
       - WOODPECKER_GITEA_SECRET=${WOODPECKER_GITHUB_SECRET}

  gitea-database:
    image: postgres:12.4-alpine
    environment:
      POSTGRES_USER: gitea
      POSTGRES_PASSWORD: 123456
      POSTGRES_DB: gitea
      PGDATA: /var/lib/postgresql/data/pgdata
    volumes:
      - pgsql:/var/lib/postgresql/data/pgdata

  gitea:
    image: gitea/gitea:1.21
    ports:
      - 3000:3000
    volumes:
      - gitea:/data
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
    depends_on:
      - gitea-database
    environment:
      USER_UID: 1000
      USER_GID: 1000
      # GITEA__server__DOMAIN: gitea.local.self
      GITEA__server__ROOT_URL: http://gitea:3000
      GITEA__database__DB_TYPE: postgres
      GITEA__database__HOST: gitea-database:5432
      GITEA__database__NAME: gitea
      GITEA__database__USER: gitea
      GITEA__database__PASSWD: 123456
      GITEA__webhook__ALLOWED_HOST_LIST: '*'
    extra_hosts:
      - 'host.docker.internal:host-gateway'

volumes:
  gitea:
  pgsql:
  woodpecker-server-data:
```

anbraten commented on providers/scaleway/v1/config.go:
```diff
@@ -0,0 +1,91 @@
+package v1
```
Would actually keep things simple and place this directly in `providers/scaleway`

raskyld replied:
I can avoid versioning the package, but it relies on API and API are versioned, but yeah we can handle it later in another way

anbraten commented on providers/scaleway/v1/flags.go:
```diff
@@ -0,0 +1,179 @@
+		Usage:    "How much storage to provision for your agents in bytes",
+		EnvVars:  []string{envPrefix + "_STORAGE_SIZE"},
+		Category: category,
+		Value:    25000000000,
```
Probably better to multiply internally and use GB instead, I guess no one gets the amount of zeros right the first time when looking at this.

anbraten commented on providers/scaleway/v1/flags.go:
```diff
@@ -0,0 +1,179 @@
+		Category: category,
+		Value:    25000000000,
+	},
+	&cli.IntFlag{
+		Name:     flagPrefix + "-client-max-retries",
+		Usage:    "How much times should we retry requests (< 0: infinite, 0: no retry)",
+		EnvVars:  []string{envPrefix + "_CLIENT_MAX_RETRIES"},
+		Category: category,
+		Value:    5,
+	},
```
Is the whole backoff logic really needed. Looks a bit overengineered to me for the moment. If testing shows this is really a thing I would rather add it to the general autoscaler logic.

raskyld replied:
I am fine removing the back-off logic from the provider, but network failures happens every time so in the future we should at least consider adding it to the general autoscaler logic like you suggest

anbraten replied:
> I am fine removing the back-off logic from the provider, but network failures happens every time so in the future we should at least consider adding it to the general autoscaler logic like you suggest

Sure that could happen, but I would suggest to follow a minimalist development approach. There is always a lot that could potentially happen so lets fix those things as soon as they arise.

raskyld commented:
Thanks for you review, I will try to fix it tonight!

raskyld commented:
I addressed your concerns, my colleagues gave me one or two tips like using the `terminate` action instead of deletion to also detach and remove the block storage but I will probably do it in another PR to avoid putting too much in this one :)

6543 commented:
could you merge master into the feature branch :)

runephilosof-karnovgroup commented on providers/scaleway/provider.go:
```diff
@@ -0,0 +1,255 @@
+		Content:  bytes.NewBufferString(ud),
+	}
+
+	err = api.SetServerUserData(&req, scw.WithContext(ctx))
```
This should be secure enough. :heavy_check_mark: 
I was wondering if this was secure, because on some clouds (Google), the user data (called metadata on Google) is accessible to all users on the instance, by just making a curl call.
https://www.scaleway.com/en/developers/api/instance/
> From a running Instance, by using the metadata API at http://169.254.42.42/user_data. To enhance security, we only allow user data viewing and editing as root. To know if the query is issued by the root user, we only accept queries made from a local port below 1024 (by default, non-root users can not bind ports below 1024). To specify the local port with cURL, use curl --local-port 1-1024 http://169.254.42.42/user_data

xoxys commented:
@raskyld Could you please rebase again?

raskyld commented:
Done :)
Sorry for the delay!

xoxys commented:
No worries. Thanks!

xoxys commented on cmd/woodpecker-autoscaler/main.go:
```diff
@@ -25,6 +26,13 @@ func setupProvider(ctx *cli.Context, config *config.Config) (engine.Provider, er
 	// Enable it again when the issue is fixed.
 	// case "linode":
 	// 	return linode.New(ctx, config)
+	case "scaleway":
+		scwCfg, err := scaleway.FromCLI(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		return scaleway.New(scwCfg, config)
```
Was there a specific reason to use another structure to do the config setup for this provider? I would prefer to keep the setup for providers equal. Everything done by ` scaleway.FromCLI(ctx)` should be handled by `scaleway.New` similar to e.g. https://github.com/woodpecker-ci/autoscaler/blob/main/providers/hetznercloud/provider.go#L43

raskyld replied:
I think I initially decoupled the parsing from CLI and the config structure to allow using declarative file (e.g. a yaml to config the autoscaler) instead of the CLI. But it seems like we won't support configuration from file so I can remove that indeed.

xoxys replied:
If you have some time to change it that would be great. If not we can also merge the current state and change it later.

xoxys commented on providers/scaleway/doc.go:
```diff
@@ -0,0 +1,14 @@
+// Package scaleway implements a way to use the Scaleway Cloud Provider for your
```
Not sure about this one. As we don't use such a pattern anywhere in the woodpecker org, I would prefer to remove it. @anbraten do you have an opinion regarding this?

raskyld replied:
I can remove it, but that's a standard used in Go to centralise the `overview` documentation of a package in a code unit.

e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/doc.go
gives https://pkg.go.dev/k8s.io/kubernetes/pkg/kubelet#pkg-overview

xoxys replied:
Im ok with it if nobody else has another opinion.

pat-s commented:
@xoxys @raskyld @6543 What is the current state here?

xoxys commented:
See my review.
